File: backend/routes/bookings/ticket/ticketGenerator.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');
const QRCode = require('qrcode');


let browserPromise = (() => {
    // Detect the platform and set appropriate options
    const os = require('os');
    const platform = os.platform();
    
    let launchOptions = {
        headless: true,
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    };
    
    // Platform-specific browser configuration
    if (platform === 'linux') {
        // For Linux (Docker environments)
        const fs = require('fs');
        if (fs.existsSync('/usr/bin/chromium-browser')) {
            launchOptions.executablePath = '/usr/bin/chromium-browser';
        }
    } else if (platform === 'darwin') {
        // For macOS - try Chrome if bundled Chromium fails
        const fs = require('fs');
        const chromePath = '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';
        if (fs.existsSync(chromePath)) {
            // Let Puppeteer use bundled Chromium first, fallback to Chrome if needed
            // Removing executablePath to let Puppeteer handle it automatically
        }
    }
    // For Windows and other platforms, let Puppeteer find the browser automatically
    
    return puppeteer.launch(launchOptions);
})();


async function generateTicketPDF(ticketData) {
    // Load your HTML template (could use a template engine for dynamic data)
    const templatePath = path.join(__dirname, 'template.html');

    let html = fs.readFileSync(templatePath, 'utf8');

    // generate QR code image
    let qrCodeDataURL;
    if (ticketData.tickets && ticketData.tickets.code) {
        try {
            qrCodeDataURL = await QRCode.toDataURL(ticketData.tickets.code, {
                errorCorrectionLevel: 'H',
                type: 'image/png',
                width: 300,
                margin: 1,
            });
        }
        catch (error) {
            console.error('Error generating QR code:', error);
            throw new Error('Failed to generate QR code');
        }
    }

    // Replace placeholders with actual ticket data
    html = html.replace(/{{name}}/g, ticketData.trips.users.name)
               .replace(/{{flight}}/g, ticketData.tickets.flights.code)
               .replace(/{{date}}/g, ticketData.localDepartureDate)
               .replace(/{{from}}/g, ticketData.tickets.flights.routes.airports_routes_departureToairports.name)
               .replace(/{{from_code}}/g, ticketData.tickets.flights.routes.airports_routes_departureToairports.id.toString())
               .replace(/{{to}}/g, ticketData.tickets.flights.routes.airports_routes_destinationToairports.name)
               .replace(/{{to_code}}/g, ticketData.tickets.flights.routes.airports_routes_destinationToairports.id.toString())
               .replace(/{{seat}}/g, ticketData.seats.postion)
               .replace(/{{class}}/g, ticketData.tickets.type)
               .replace(/{{ticket_code}}/g, ticketData.tickets.code)
               .replace(/{{airline}}/g, ticketData.tickets.flights.airline_name)
               .replace(/{{barcodeIMG}}/g, qrCodeDataURL)
               .replace(/{{extras_description}}/g, ticketData.extras ? ticketData.extras.description : '')
               .replace(/{{extras_price}}/g, ticketData.extras && ticketData.extras.price !== undefined ? ticketData.extras.price + ' €' : 'No extras included');

    // Launch Puppeteer and generate PDF
    try {
        const browser = await browserPromise; // Reuse the same browser
        const page = await browser.newPage();
        await page.setContent(html, {
            timeout: 0,
            waitUntil: 'domcontentloaded',
        });
        await page.emulateMediaType('print');
        const pdfBuffer = await page.pdf({
            preferCSSPageSize: true,
            timeout: 0,
            format: 'A4'
        });
        await page.close();
        return pdfBuffer;
    } catch (error) {
        console.error('Error generating PDF:', error);
        throw error;
    }
}

module.exports = { generateTicketPDF };

File: backend/routes/bookings/ticket/ticketGenerator.py
import os
import asyncio
import qrcode
import base64
from pyppeteer import launch
from pathlib import Path
from io import BytesIO

TEMPLATE_PATH = Path(__file__).parent / 'template.html'

_browser = None

async def get_browser():
    global _browser
    if _browser is not None:
        return _browser
    platform = os.uname().sysname.lower()
    launch_options = {
        'headless': True,
        'args': ['--no-sandbox', '--disable-setuid-sandbox'],
        'executablePath': '/usr/bin/chromium'  # o '/usr/bin/chromium-browser'
    }
    if platform == 'linux':
        if Path('/usr/bin/chromium-browser').exists():
            launch_options['executablePath'] = '/usr/bin/chromium-browser'
    _browser = await launch(**launch_options)
    return _browser

def generate_qr_code(data: str) -> str:
    qr = qrcode.QRCode(
        error_correction=qrcode.constants.ERROR_CORRECT_H,
        box_size=10,
        border=1,
    )
    qr.add_data(data)
    qr.make(fit=True)
    img = qr.make_image(fill_color="black", back_color="white")
    buf = BytesIO()
    img.save(buf, format='PNG')
    b64 = base64.b64encode(buf.getvalue()).decode('utf-8')
    return f'data:image/png;base64,{b64}'

def fill_template(html: str, ticket_data: dict, qr_code_data_url: str) -> str:
    extras = ticket_data.get('extras', {})
    html = html.replace('{{name}}', ticket_data['trips']['users']['name'])
    html = html.replace('{{flight}}', ticket_data['tickets']['flights']['code'])
    html = html.replace('{{date}}', ticket_data['localDepartureDate'])
    html = html.replace('{{from}}', ticket_data['tickets']['flights']['routes']['airports_routes_departureToairports']['name'])
    html = html.replace('{{from_code}}', str(ticket_data['tickets']['flights']['routes']['airports_routes_departureToairports']['id']))
    html = html.replace('{{to}}', ticket_data['tickets']['flights']['routes']['airports_routes_destinationToairports']['name'])
    html = html.replace('{{to_code}}', str(ticket_data['tickets']['flights']['routes']['airports_routes_destinationToairports']['id']))
    html = html.replace('{{seat}}', ticket_data['seats']['postion'])
    html = html.replace('{{class}}', ticket_data['tickets']['type'])
    html = html.replace('{{ticket_code}}', ticket_data['tickets']['code'])
    html = html.replace('{{airline}}', ticket_data['tickets']['flights']['airline_name'])
    html = html.replace('{{barcodeIMG}}', qr_code_data_url)
    html = html.replace('{{extras_description}}', extras.get('description', ''))
    html = html.replace('{{extras_price}}', f"{extras.get('price', 'No extras included')} €" if 'price' in extras else 'No extras included')
    return html

async def generate_ticket_pdf(ticket_data: dict) -> bytes:
    html = TEMPLATE_PATH.read_text(encoding='utf-8')
    qr_code_data_url = ''
    if ticket_data.get('tickets', {}).get('code'):
        try:
            qr_code_data_url = generate_qr_code(ticket_data['tickets']['code'])
        except Exception as error:
            print('Error generating QR code:', error)
            raise RuntimeError('Failed to generate QR code')
    html = fill_template(html, ticket_data, qr_code_data_url)
    try:
        browser = await get_browser()
        page = await browser.newPage()
        await page.setContent(html)
        await page.emulateMedia('print')
        pdf_bytes = await page.pdf(preferCSSPageSize=True, timeout=0, format='A4')
        await page.close()
        return pdf_bytes
    except Exception as error:
        print('Error generating PDF:', error)
        raise

# Usage example (async context):
# pdf = asyncio.run(generate_ticket_pdf(ticket_data))